### PR TITLE
nextcloud20: init

### PR DESCRIPTION
**Pull request by Ma27**
###### Motivation for this change

Adds [`nextcloud20`](https://github.com/nextcloud/server/releases/tag/v20.0.0) to our package-set. For this change, a few customizations were needed. Please review this commit-by-commit for more details.

To allow users to upgrade to the latest major (even on stable NixOS), the addition of the package (and the `nginx` changes) **should be backported** to 20.09 (cc @NixOS/backports ), the changes to the upgrade-path should be left out though[1].

[1] https://nixos.org/manual/nixos/stable/index.html#module-services-nextcloud-maintainer-info

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-apps/nextcloud.nix [hunk 1]
```diff
@@ -330,37 +330,27 @@ in {
...
-        ++ (optional (versionOlder cfg.package.version "18") ''
-          A legacy Nextcloud install (from before NixOS 20.03) may be installed.
-
-          You're currently deploying an older version of Nextcloud. This may be needed
-          since Nextcloud doesn't allow major version upgrades that skip multiple
-          versions (i.e. an upgrade from 16 is possible to 17, but not 16 to 18).
-
-          It is assumed that Nextcloud will be upgraded from version 16 to 17.
+      warnings = let
+        upgradeWarning = major: latest: nixos:
```
Comment (lheckemann) [inline nixos/modules/services/web-apps/nextcloud.nix:334]: Does `latest` need to be a parameter here? It doesn't seem to be used with anything other than `20`?

**Review by lheckemann — APPROVED** [review]
LGTM, just a nit.

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 1 above)
Comment (mweinelt) [reply nixos/modules/services/web-apps/nextcloud.nix:334]: I think if we have `latest` that `major` would be better understandable as `current`.

Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix:334]: @mweinelt what do you mean?

Comment (lheckemann) [reply nixos/modules/services/web-apps/nextcloud.nix:334]: AFAIU, only _if_ we keep the `latest` parameter, rename the `major` parameter to `current`. I'm for dropping the `latest` parameter though :)

Comment (mweinelt) [reply nixos/modules/services/web-apps/nextcloud.nix:334]: I have no strong opinion on this anyhow. The rest of the PR looks good from a cursory glance.

**Review by mweinelt — APPROVED** [review]

File: nixos/modules/services/web-apps/nextcloud.nix (hunk 1 above)
Comment (Ma27) [reply nixos/modules/services/web-apps/nextcloud.nix]: @lheckemann fixed :)

**Comment by Ma27** [comment]
Ported to 20.09 (to make sure that people can upgrade to latest nextcloud whenever they want) excluding the upgrade-path patch: 8597315fae509af13b30ead95fdc1900d5530fa5, 452603b385cb72a020f4e45ff2d76d10cc664b9a, 555375859b76416ffe42feecef738a4d153cc7dc

**Pull request merged**